File: leetcode/c++/StrStr.cpp
#include<iostream>
#include<string>

using namespace std;

typedef struct Test {
  string haystack;
  string needle;
  int exp;
} Test;

const Test tests[] = {
	{"sadbutsad", "sad", 0},
	{"leetcode", "leeto", -1},
	{"a", "a", 0},
	{"ba", "a", 1},
};

int strStr(const string haystack, const string needle) {
  if (needle.size() == 0 || needle.size() > haystack.size())
    return -1;
  const int n = haystack.size(), m = needle.size();
  for (int i = 0; i < n-m+1; i++)
    if (haystack.substr(i, m) == needle)
      return i;
  return -1;
}

// Much easier solution
int _strStr2(const string haystack, const string needle) {
  return haystack.find(needle);
}

int main() {
  cout << "Running..." << endl;
  for (const auto t : tests) {
    const auto res = strStr(t.haystack, t.needle);
    if (res == t.exp) {
      cout << "Success" << endl;
    } else {
      cout << "Failure: " << res << " != " << t.exp << endl;
    }
  }
  return 0;
}

File: leetcode/rust/src/str_str.rs

fn str_str(haystack: String, needle: String) -> i32 {
  if needle.len() == 0 || needle.len() > haystack.len() {
    return -1;
  }

  let n = haystack.len();
  let m = needle.len();
  for i in 0..n-m+1 {
    if haystack[i..i+m] == needle {
      return i as i32;
    }
  }

  return -1
}

// Another, more memory-efficient solution from LeetCode
fn _str_str2(haystack: String, needle: String) -> i32 {
  let mut step = 0;
  loop {
    if step + needle.len() > haystack.len() {
      return -1i32
    }
    let check = &haystack[step..needle.len() + step].to_string();
    if check.eq(&needle) {
      return step as i32
    }
    step += 1;
  }
}

// Another, weirder solution
fn _str_str3(haystack: String, needle: String) -> i32 {
  for (i, s) in haystack
    .chars()
    .collect::<Vec<char>>()
    .windows(needle.len())
    .enumerate()
  {
    if s.into_iter().collect::<String>() == needle {
      return i as i32;
    }
  }
  return -1;
}


#[cfg(test)]
pub mod tests {
  use super::*;

  #[test]
  fn test1() { 
    assert_eq!(str_str("sadbutsad".to_string(), "sad".to_string()), 0);
  }

  #[test]
  fn test2() { 
    assert_eq!(str_str("leetcode".to_string(), "leeto".to_string()), -1);
  }

  #[test]
  fn test3() { 
    assert_eq!(str_str("a".to_string(), "a".to_string()), 0);
  }

  #[test]
  fn test4() { 
    assert_eq!(str_str("ba".to_string(), "a".to_string()), 1);
  }
}
